[2.12] Improve -Yrepl-class-based

dwijnand opened the pull request:
These changes improve the REPL under `-Yrepl-class-based`, so that it eventually it will become the new default (in 2.13.2).

For background: `-Yrepl-class-based` is an experimental (`-Y`) compiler option that switches the REPL to use "class-based" (instead of "object-based") wrappers, as its implementation detail.  This mode originated in Spark's fork of the Scala REPL (and was later contributed upstream to `scala/scala`) as the object-based wrappers were problematic for the code serialisation that the Spark REPL does.

Additionally, the change in encoding to Scala's lambdas that happened in Scala 2.12 (i.e. since targeting JDK 8+) made it much, much easier to cause a deadlock in the REPL, aka scala/bug#9076.  Using class-based wrappers resolves this too.

However, there were some limitations to the class-based wrappers' implementation, that have been addressed with this PR.

1. The first is the most involved part of this change.  The mode contained a heuristic that an import from previous lines would be ignored if the current line contained a class definition `C` and the import didn't contribute a name that was among the identifiers on this line. The goal of this heuristic was to avoid the outer pointer of class `C` transitively capturing the `iw$` wrappers with state from previous lines that wasn't actually referred to.  However, these imports could contribute implicits which are needed during typechecking of the class, so the heuristic changed semantics w.r.t `-Yrepl-class-based:false` (reported as [SPARK-5150][]).  This PR always considers the imports, but adds a post-processing step to prune out unused references to the state of prior lines. 

2. The mode can now handle value classes, by switching the evaluation to use the old, object-based wrappers for the snippet that defines a value class.

3. The mode now mitigates any attempts to define a macro in the REPL, by giving the user some more information in the error message.

Refs scala/bug#9076
Refs [SPARK-5150][]

[SPARK-5150]: https://issues.apache.org/jira/browse/SPARK-5150

TODOs:
* [x] review the changes bearing in mind Spark uses `-Yrepl-class-based` already

retronym committed:
Clean unused temp vals in REPL wrappers and remove heuristic that ignores some imports

Co-authored-by: Dale Wijnand <dale.wijnand@gmail.com>

dwijnand committed:
repl: Always use .INSTANCE accessor

Always generate a `val INSTANCE` field, whether it's for class-based or
object-based, so that the access code branches needn't branch on
isClassBased.

This greatly simplifies the next commit, for handling value classes.

dwijnand committed:
repl: Handle value classes in class-based

... by making any code snippet that defines (or seems to define) a value
class use the original "object-based" wrappers instead.

retronym committed:
repl: Error nicely on macros in class-based

Co-authored-by: Dale Wijnand <dale.wijnand@gmail.com>

dwijnand committed:
Switch unusedPrivate to not need private in REPL

This avoids problems like

    private value read escapes its defining scope as part of type u.Type

when running `test/files/run/class-symbol-contravariant.scala` (and a
few others) with -Yrepl-class-based.

Also add a test specific for type aliases, extracted from
`test/files/jvm/interpreter.scala`, to run in class-based mode.

dwijnand committed:
Default -Yrepl-class-based to true

... and thus -Yrepl-use-magic-imports too.

dwijnand committed:
Default false -Yrepl-{class-based,use-magic-imports}

The intent is for these to be true, but we'll do that for 2.13.x,
not 2.12.x.

som-snytt commented on src/repl/scala/tools/nsc/interpreter/ReplGlobal.scala:
```diff
@@ -40,6 +42,42 @@ trait ReplGlobal extends Global {
 
+  override protected def computeInternalPhases(): Unit = {
+    super.computeInternalPhases()
+    addToPhasesSet(wrapperCleanup, "Remove usused values from import wrappers to avoid unwanted capture of non-serializable objects")
```
unused

som-snytt commented on src/repl/scala/tools/nsc/interpreter/ReplGlobal.scala:
```diff
@@ -14,13 +14,15 @@ package scala.tools.nsc
 import typechecker.Analyzer
 
 /** A layer on top of Global so I can guarantee some extra
  *  functionality for the repl.  It doesn't do much yet.
```
Maybe not say that any more.

som-snytt commented on test/files/run/macro-bundle-repl.scala:
```diff
@@ -1,6 +1,12 @@
+    override def transformSettings(s: Settings) = {
+    s.Yreplclassbased.value = false // macros are object-based only
+    s
+  }
```
`s.tap(_.Yreplthing = false)`. Use it or lose it.

dwijnand replied:
Good idea.  I'll switch to that on 2.13.x

som-snytt commented on test/files/run/repl-colon-type.check:
```diff
@@ -39,7 +39,7 @@ scala> :type protected lazy val f = 5
  object $iw where target is defined
          lazy val $result = f
-                                                  ^
+                                                           ^
```
ghostly! like a poltergeist moving a chair.

som-snytt commented on test/files/run/repl-trim-stack-trace.check:
```diff
@@ -6,23 +6,23 @@ f: Nothing
 scala> f
 java.lang.Exception: Uh-oh
-  at .f(<console>:XX)
+  at f(<console>:11)
```
IIRC this sort of thing was to avoid spurious diffs.

som-snytt reviewed (commented):
Gave it a glance.

retronym commented:
In https://github.com/retronym/spark/pull/1, I've setup the Spark Maven build to be able to test against the Scala release from this (or any) PR, figured out how to run the Spark REPL test suite, and found that we need to add `-Yrepl-use-magic-imports` to the Spark REPL setup to keep the test green.

retronym commented:
I'd prefer if we didn't require Spark to pass `-Ymagic-imports`. (That would require a new Spark release.)

We could enable it by default if `-Yrepl-class-based` is set, expanding the help of that option to say we'll do this. We do this sort of thing elsewhere using `Setting#withPostSetHook`.

retronym commented:
/cc @srowen (from the Spark team)

lrytz commented:
> we need to add `-Yrepl-use-magic-imports` to the Spark REPL setup to keep the test green.

Do you know why, what change "broke" plain `-Yrepl-class-based` for spark?

dwijnand commented:
> [dwijnand](/dwijnand) modified the milestones: [2.12.12](/scala/scala/milestone/85), [2.12.11](/scala/scala/milestone/84) [now](#event-3034171819)

... optimistically.

dwijnand committed:
Make -Yrepl-class-based set -Yrepl-use-magic-imports

The Spark REPL uses -Yrepl-class-based and the change around
ImportHandler in interpreter/Imports (and that introduced the custom
wrapperCleanup phase for the REPL) requires -Yrepl-use-magic-imports be
set in order for the Spark REPL to continue to work correctly.  Rather
than affect all existing Spark releases that choose to upgrade Scala, we
preserve the functioning of the Spark REPL by enabling "magic imports"
when using the class-based wrappers.

retronym commented:
> Do you know why, what change "broke" plain `-Yrepl-class-based` for spark?

It will be the [removal](https://github.com/scala/scala/pull/8712/files#diff-57ebae7a4267bf7a735609cfb18cf8b6L141) of the heuristic that ignores imports introduced by previous lines for when the current line contains a class definitions that doesn't include an identifier contributed by that import.

That was supposed to be okay for us to remove because the new dead back-reference elimination (DBRE) should detect that its unused and kill it.

The use case is covered in `run/repl-serialization.scala`. I see now that I added `-Yrepl-use-magic-imports` to that test, so there must have been an issue with making the DBRE work with classic wrapper nesting. 

Indeed, that test fails with `-Yrepl-class-based -Yrepl-use-magic-imports:false`:

```diff
-	at scala.tools.partest.nest.StreamCapture$.$anonfun$capturingOutErr$2(StreamCapture.scala:32)
diff --git a/test/files/run/repl-serialization.scala b/test/files/run/repl-serialization.scala
index a0fed0b6f4..2f201146c2 100644
--- a/test/files/run/repl-serialization.scala
+++ b/test/files/run/repl-serialization.scala
@@ -14,7 +14,7 @@ object Test {
   def run(): Unit = {
     val settings = new Settings()
     settings.Yreplclassbased.value = true
-    settings.YreplMagicImport.value = true
+    settings.YreplMagicImport.value = false
     //settings.Xprint.value = List("refchecks","wrapper-cleanup")
     settings.usejavacp.value = true
```

```
# partest /Users/jz/code/scala/test/files/run/repl-serialization.scala
% scalac repl-serialization.scala
% <in process execution of run/repl-serialization.scala> > repl-serialization-run.log
% diff /Users/jz/code/scala/test/files/run/repl-serialization-run.log /Users/jz/code/scala/test/files/run/repl-serialization.check
@@ -23,75 +23,5 @@ u: U = U
   evaluating O
   constructing A
 == reconstituting into a fresh classloader
-java.io.NotSerializableException: $line9.$read$$iw$$iw$TestClass
-	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
```

I'll take a quick look and see if the if its an easy or hard problem to solve.

retronym commented:
## `-Yrepl-use-magic-imports:true`

```
package $line17 {
  sealed class $read extends _root_.java.io.Serializable {
    val $line2$read: $line2.$read.INSTANCE.type = $line2.$read.INSTANCE();
    val $line3$read: $line3.$read.INSTANCE.type = $line3.$read.INSTANCE();
    val $line4$read: $line4.$read.INSTANCE.type = $line4.$read.INSTANCE();
    val $line6$read: $line6.$read.INSTANCE.type = $line6.$read.INSTANCE();
    val $line7$read: $line7.$read.INSTANCE.type = $line7.$read.INSTANCE();
    val $line8$read: $line8.$read.INSTANCE.type = $line8.$read.INSTANCE();
    val $line9$read: $line9.$read.INSTANCE.type = $line9.$read.INSTANCE();
    val $line15$read: $line15.$read.INSTANCE.type = $line15.$read.INSTANCE();
    val $line16$read: $line16.$read.INSTANCE.type = $line16.$read.INSTANCE();
    sealed class $iw extends _root_.java.io.Serializable {
      val res0 = $read.this.$line2$read().$iw().extract().apply({
        def $anonfun$res0(): A = new A($read.this.$line3$read().$iw().x().+($read.this.$line4$read().$iw().getX()).+($read.this.$line6$read().$iw().y()).+($read.this.$line7$read().$iw().z()).+($read.this.$line7$read().$iw().zz()).+($read.this.$line8$read().$iw().O().apply()).+($read.this.$line16$read().$iw().u().x()));
        (() => $anonfun$res0())
      })
    };
    val $iw = new $iw()
  }

  object $read extends Serializable {
    val INSTANCE = new $line17.$read()
  }
}
```

## `-Yrepl-use-magic-imports:true`

```
package $line17 {
  sealed class $read extends _root_.java.io.Serializable {
    sealed class $iw extends _root_.java.io.Serializable {
      val $line2$read: $line2.$read.INSTANCE.type = $line2.$read.INSTANCE();
      val $line3$read: $line3.$read.INSTANCE.type = $line3.$read.INSTANCE();
      val $line4$read: $line4.$read.INSTANCE.type = $line4.$read.INSTANCE();
      val $line6$read: $line6.$read.INSTANCE.type = $line6.$read.INSTANCE();
      val $line7$read: $line7.$read.INSTANCE.type = $line7.$read.INSTANCE();
      val $line8$read: $line8.$read.INSTANCE.type = $line8.$read.INSTANCE();
      val $line9$read: $line9.$read.INSTANCE.type = $line9.$read.INSTANCE();
      sealed class $iw extends _root_.java.io.Serializable {
        sealed class $iw extends _root_.java.io.Serializable {
          val $line15$read: $line15.$read.INSTANCE.type = $line15.$read.INSTANCE();
          val $line16$read: $line16.$read.INSTANCE.type = $line16.$read.INSTANCE();
          sealed class $iw extends _root_.java.io.Serializable {
            val res0 = $iw.this.$line2$read().$iw().$iw().extract().apply({
              def $anonfun$res0(): A = new A($iw.this.$line3$read().$iw().$iw().x().+($iw.this.$line4$read().$iw().$iw().getX()).+($iw.this.$line6$read().$iw().$iw().y()).+($iw.this.$line7$read().$iw().$iw().z()).+($iw.this.$line7$read().$iw().$iw().zz()).+($iw.this.$line8$read().$iw().$iw().O().apply()).+($iw.this.$line16$read().$iw().$iw().$iw().$iw().u().x()));
              (() => $anonfun$res0())
            })
          };
          val $iw = new $iw()
        };
        val $iw = new $iw()
      };
      val $iw = new $iw()
    };
    val $iw = new $iw()
  }

  object $read extends Serializable {
    val INSTANCE = new $line17.$read()
  }
}
```

retronym committed:
Make REPL dead backreference elimination work with nested wrapped

The custom phase that does this needs to descend deeper than
the first `Template`.

It now works bottom up (from the innermost `$iw`) and runs the
unused analysis at each level, feeding the possibly simpilfied
trees into the analysis of the enclosing level.

I also needed to add a special case for the `lineXY$read`
variable names which should be considered private for the
purpose of this analysis.

retronym committed:
Revert to using magic imports in REPL serialization test

retronym commented:
I pushed a fix. The main problem was that the `wrapper-cleanup` phase was not recursing beyond the outermost `ClassDef`.

dwijnand commented:
```
# Failed test paths (this command will update checkfiles)
partest --update-check \
  /home/travis/build/scala/scala/test/files/run/repl-no-imports-no-predef.scala \
  /home/travis/build/scala/scala/test/files/run/repl-paste-4.scala \
  /home/travis/build/scala/scala/test/files/run/repl-paste.scala
```

retronym commented:
I've pushed a small refinement -- we should not run the unused analysis on the each `ClassDef` `ModuleDef` defined in a line separately, we need to first run it on the innermost `class $iw`.

retronym committed:
Be more selective in REPL dead backreference elimination

  - Only run the transform under `-Yrepl-class-based`
  - Only eliminate the `$lineXY$read` fields and accessors to be sure
    we don't accidentally eliminate statically unused user code,
    which could be intended for reflective use or the like.
  - Colocate the logic that creates and tests for `$lineXY$read`
    val names.

retronym commented:
I also pushed a commit to limit the unsed private removal to `-Yrepl-class-based`. I have some further WIP that I think we also need to add (I'll get it working and submit by Monday): https://github.com/retronym/scala/tree/repl/tighten-unused-wip

That last part aims to centralize the logic for creating the REPL temp val names and checking if symbol has such a name. It then aims to limit the unused private removal to such vals, as we don't have any business removing unsed privates from user code.

retronym commented:
```
./build/mvn -P scala-pr-validation-snapshots -Dscala.version=2.12.11-bin-1dc6d84-SNAPSHOT -pl :spark-repl_2.12 -Dtest=none -DwildcardSuites='org.apache.spark.repl.ReplSuite,org.apache.spark.repl.SingletonReplSuite,org.apache.spark.repl.ExecutorClassLoaderSuite' test
...
All tests passed.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:09 min
[INFO] Finished at: 2020-02-17T07:28:33+10:00
[INFO] ------------------------------------------------------------------------
```

retronym commented:
Also passes with:

```
index a1f59a0a2e..baea3ebef6 100644
--- a/repl/src/main/scala/org/apache/spark/repl/Main.scala
+++ b/repl/src/main/scala/org/apache/spark/repl/Main.scala
@@ -65,41 +65,41 @@ object Main extends Logging {
       // Remove file:///, file:// or file:/ scheme if exists for each jar
       .map { x => if (x.startsWith("file:")) new File(new URI(x)).getPath else x }
       .mkString(File.pathSeparator)
     val interpArguments = List(
       "-Yrepl-class-based",
       "-Yrepl-outdir", s"${outputDir.getAbsolutePath}",
       "-classpath", jars
     ) ++ args.toList

     val settings = new GenericRunnerSettings(scalaOptionError)
     settings.processArguments(interpArguments, true)

     // Support Scala 2.12.10 and 2.12.11. The latter introduces `-Yrepl-use-magic-imports` and Spark
     // requires that to be enabled to take advantage of the dead code eliminiation now in `-Yclass-based`
     // that solves https://issues.apache.org/jira/browse/SPARK-5150.
     //
     // See https://github.com/scala/scala/pull/8576
     // See https://github.com/scala/scala/pull/8712
     //
     // TODO: Move the setting into the list above once support for 2.12.10 and below is dropped.
-    settings.processArguments("-Yrepl-use-magic-imports" :: Nil, processAll = false)
+    settings.processArguments("-Yrepl-use-magic-imports:false" :: Nil, processAll = false)
```

retronym committed:
Simplfy wrapper-cleanup REPL phase

I'lm pretty sure we no longer need to customised `ignoreNames` now
that we're customizing `isEffectivelyPrivate` and we're restricting
the unused decl dropping to trees of the form `val $lineXY$read = ...`

lrytz commented on src/repl/scala/tools/nsc/interpreter/ReplGlobal.scala:
```diff
@@ -40,6 +42,35 @@ trait ReplGlobal extends Global {
+    override val global: self.type = self
+    override val phaseName: String = "wrapper-cleanup"
+    /** Names of phases that must run before this phase. */
+    override val runsAfter: List[String] = List("refchecks")
```
Should we also define `runsBefore` to avoid surprises? (https://github.com/scala/scala/pull/8426)

lrytz commented on src/repl/scala/tools/nsc/interpreter/MemberHandlers.scala:
```diff
@@ -205,6 +206,10 @@ trait MemberHandlers {
     override def definesTerm = Some(name.toTermName) filter (_ => mods.isCase)
+    override final def definesValueClass: Boolean = member.impl.parents match {
+      case Ident(tpnme.AnyVal) :: _ => true // approximating with a syntactic check
+      case _                        => false
```
should we recurse into the `impl`?

```
scala> object T { class Meter(val value: Int) extends AnyVal }
<console>:11: error: value class may not be a member of another class
       object T { class Meter(val value: Int) extends AnyVal }
                        ^
```

lrytz reviewed (commented):
Looks great! Should we disable the `postSetHook` that enables magic imports?

dwijnand committed:
Opt VCs nested in modules out of class-based-repl

dwijnand committed:
Revert "Make -Yrepl-class-based set -Yrepl-use-magic-imports"

This reverts commit c5f14241d74a4f3731b357d44f9589c6521cd4fd.

lrytz commented on test/files/run/repl-serialization.scala:
```diff
@@ -14,6 +14,9 @@ object Test {
   def run(): Unit = {
     val settings = new Settings()
     settings.Yreplclassbased.value = true
+    // The postSetHook of the above implies settings.YreplMagicImport.value = true
```
no longer tue - do we need `settings.YreplMagicImport.value = true` now?

dwijnand replied:
No, I think not.  I think the point is that Jason's "REPL dead backreference elimination" works with and without.  So it should be tested under `settings.Yreplclassbased.value = true` (what Spark uses) and the default `YreplMagicImport`.  I'll drop the comment.

retronym reviewed (approved):
:shipit:

dwijnand committed:
repl: Nail the wrapper-cleanup phase ordering

This should help ensure the order is as deterministic as possible.

SethTisue commented:
@dwijnand tentatively labeled with "release-notes", but actually I'm unsure what the takeaway here is for users. I'd suggest either removing the label, or improving the PR description to begin with a user-focused description, with the contributor-focused info afterwards.

srowen commented:
For Spark, is the takeaway that `spark-shell` needs to set `-Yrepl-use-magic-imports` if Spark updates the Scala 2.12 version?

dwijnand commented:
@srowen No, that's no longer true, thankfully (it's why we were able to drop the postSetHook).

The Spark-specific takeaway is we tweaked how to minimise references, so that Spark's serialisation continues to work (aka not regressing on https://issues.apache.org/jira/browse/SPARK-2576 and https://issues.apache.org/jira/browse/SPARK-2632), while fixing https://issues.apache.org/jira/browse/SPARK-5150.  Tested with our `test/files/run/repl-serialization.scala`, as well as running `org.apache.spark.repl.ReplSuite`, `org.apache.spark.repl.SingletonReplSuite`, and `org.apache.spark.repl.ExecutorClassLoaderSuite` in the Spark build (though, admittedly, not on the latest sha).

If you could double-check that for us, using the https://scala-ci.typesafe.com/artifactory/scala-integration/org/scala-lang/scala-compiler/2.12.11-bin-4c726db/ that would be great!

dwijnand commented:
@SethTisue I'll update the description.  It won't be interesting for non-Spark users until 2.13.2, but we can still describe it here and reference to it.
